### PR TITLE
feat(snql): SnQL Grammar and Visitor, Part 3

### PR DESCRIPTION
**Pull request by ayirr7**
Updates: 

- Added support for expressions (arithmetic expressions and conditions/combinations of conditions) in parentheses
- Modified existing test cases to exhibit changes
- Fixed type hints

### CODE REVIEW

**Comment by fpacifici** [comment]
Did you actually forked the `feat/snqlvisitorp3` branch out of `feat/snqlvisitorp2` or out of master ? If I change the base of this PR to `feat/snqlvisitorp2` you see conflicts.

File: snuba/query/snql/__init__.py [hunk 1]
```diff
@@ -82,26 +97,72 @@
 )
 
 
+class AndTuple(NamedTuple):
```
Comment (evanh) [inline snuba/query/snql/__init__.py:100]: Is there a reason you need separate classes for And and Or?

File: snuba/query/snql/__init__.py [hunk 2]
```diff
@@ -29,37 +34,47 @@
...
     clause                = space* ~r"[-=><\w]+" space*
 
     low_pri_arithmetic    = space* high_pri_arithmetic space* (low_pri_tuple)*
-    high_pri_arithmetic   = space* arithmetic_term space* (high_pri_tuple)*
+    high_pri_arithmetic   = (space* arithmetic_term space* (high_pri_tuple)*)
     low_pri_tuple         = low_pri_op high_pri_arithmetic
     high_pri_tuple        = high_pri_op arithmetic_term
-    arithmetic_term       = (space*) (function_call / numeric_literal / column_name) (space*)
+
+    arithmetic_term       = (space*) (function_call / numeric_literal / column_name / interm_arithm) (space*)
```
Comment (evanh) [inline snuba/query/snql/__init__.py:75]: Why brackets on the spaces?

File: snuba/query/snql/__init__.py [hunk 3]
```diff
@@ -139,6 +200,7 @@ def visit_high_pri_arithmetic(
         node: Node,
         visited_children: Tuple[Any, Expression, Any, HighPriArithmetic],
     ) -> Expression:
+
```
Comment (evanh) [inline snuba/query/snql/__init__.py:203]: 💀

File: snuba/query/snql/__init__.py [hunk 4]
```diff
@@ -151,6 +213,63 @@ def visit_quoted_literal(
...
+        # and_condition will be an empty Node
+        if isinstance(and_condition, Node):
+            return left_condition
+        if isinstance(and_condition, (AndTuple, OrTuple)):
+            _, exp = and_condition
+            return combine_and_conditions([left_condition, exp])
+        elif isinstance(and_condition, list):
+            for elem in and_condition:
+                _, exp = elem
+                left_condition = combine_and_conditions([left_condition, exp])
```
Comment (evanh) [inline snuba/query/snql/__init__.py:243]: It feels unpythonic to call `combine_and_conditions` once for every element in the list, particularly since that function takes a list as an argument. I would build the list first, and then pass the whole thing into `combine_and_conditions`.

```
args = [left_condition] + [elem[1] for elem in and_condition]
combine_and_conditions(args)
```

File: snuba/query/snql/__init__.py [hunk 5]
```diff
@@ -394,6 +403,14 @@ def generic_visit(self, node: Node, visited_children: Any) -> Any:
         return generic_visit(node, visited_children)
 
 
+thing = SnQLVisitor().visit(
+    snql_grammar.parse(
+        "MATCH (blah) WHERE (time_seen<3 OR last_seen=afternoon) OR name=bob COLLECT a"
+    )
+)
+print(thing.get_condition_from_ast())
+
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:412]: please remove

File: snuba/query/snql/__init__.py (hunk 3 above)
Comment (fpacifici) [reply snuba/query/snql/__init__.py:203]: please address this comment.

File: snuba/query/snql/__init__.py [hunk 6]
```diff
@@ -66,7 +68,7 @@
     clause                = space* ~r"[-=><\w]+" space*
 
     low_pri_arithmetic    = space* high_pri_arithmetic space* (low_pri_tuple)*
-    high_pri_arithmetic   = space* arithmetic_term space* (high_pri_tuple)*
+    high_pri_arithmetic   = (space* arithmetic_term space* (high_pri_tuple)*)
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:71]: why do you need to put parentheses around this expression ? It seems it was right before.

File: snuba/query/snql/__init__.py [hunk 7]
```diff
@@ -46,8 +46,10 @@
     having_clause         = space* "HAVING" or_expression space*
     order_by_clause       = space* "ORDER BY" order_list space*
 
-    condition             = low_pri_arithmetic condition_op (function_call / column_name / numeric_literal / string_literal) space*
+    main_condition        = low_pri_arithmetic condition_op (function_call / column_name / numeric_literal) space*
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:49]: Why did you remove the string_literal ?

(Which by the way should be a quoted_literal)

File: snuba/query/snql/__init__.py [hunk 8]
```diff
@@ -231,14 +240,14 @@ def visit_and_expression(
         elif isinstance(and_condition, list):
             for elem in and_condition:
                 _, exp = elem
-                left_condition = combine_and_conditions([left_condition, exp])
-        return left_condition
+                args.append(exp)
+        return combine_and_conditions(args)
```
Comment (fpacifici) [inline snuba/query/snql/__init__.py:244]: thanks

File: snuba/query/snql/__init__.py (hunk 1 above)
Comment (ayirr7) [reply snuba/query/snql/__init__.py]: Yes - in visit_and_expression and visit_or_expression, having these classes is really useful because it allows me to break down the logic in those visitor functions into 3 cases. With this, there's no ambiguity and the correct Expression can be built in all 3 cases.

File: snuba/query/snql/__init__.py (hunk 2 above)
Comment (ayirr7) [reply snuba/query/snql/__init__.py]: Removed them - thanks.

**Review by fpacifici — CHANGES_REQUESTED** [review]
some comments, nothing fundamental.

File: snuba/query/snql/__init__.py (hunk 4 above)
Comment (ayirr7) [reply snuba/query/snql/__init__.py]: Good point - I changed the logic in my code so that I'm building the list first and then passing it once to combine_and_conditions / combine_or_conditions. Thank you!

File: snuba/query/snql/__init__.py (hunk 6 above)
Comment (ayirr7) [reply snuba/query/snql/__init__.py]: oops I thought both of those comments had been committed.

File: snuba/query/snql/__init__.py (hunk 7 above)
Comment (ayirr7) [reply snuba/query/snql/__init__.py]: Added quoted_literal.

File: snuba/query/snql/__init__.py (hunk 5 above)
Comment (ayirr7) [reply snuba/query/snql/__init__.py]: Removed.

**Review by fpacifici — APPROVED** [review]
excellent, thanks
